File: tagalong/021-interfaces/interfaces.py
import math
from typing import Protocol, Tuple

class Geometry(Protocol):
    def area(self) -> float: ...
    def perim(self) -> float: ...

class Rectangle:
    def __init__(self, width: float, height: float) -> None:
        self.width = width
        self.height = height
    def area(self) -> float:
        return self.width * self.height
    def perim(self) -> float:
        return 2 * (self.width + self.height)

class Circle:
    def __init__(self, radius: float) -> None:
        self.radius = radius
    def area(self) -> float:
        return math.pi * self.radius ** 2
    def perim(self) -> float:
        return 2 * math.pi * self.radius

def shape_efficiency(g: Geometry) -> float:
    return g.area() / g.perim()

def detect_circle(g: Geometry) -> Tuple[float, bool]:
    if isinstance(g, Circle):
        return g.radius, True
    return 0.0, False

if __name__ == "__main__":
    rect = Rectangle(1, 4)
    square = Rectangle(4, 4)
    c = Circle(4)

    print("1x4 rectangle efficiency:", shape_efficiency(rect))
    print("square efficiency:", shape_efficiency(square))
    print("circle efficiency:", shape_efficiency(c))

    print(detect_circle(rect))
    print(detect_circle(c))

File: tagalong/021-interfaces/interfaces.go
package main

import (
	"fmt"
	"math"
)

type geometry interface {
	area() float64
	perim() float64
}

type rectangle struct {
	width, height float64
}

type circle struct {
	radius float64
}

func (r rectangle) area() float64 {
	return r.width * r.height
}
func (r rectangle) perim() float64 {
	return 2*r.width + 2*r.height
}

func (c circle) area() float64 {
	return math.Pi * c.radius * c.radius
}
func (c circle) perim() float64 {
	return 2 * math.Pi * c.radius
}

func shapeEfficiency(g geometry) float64 {
	// Area enclosed per unit perimeter used.
	return g.area() / g.perim()
}

func detectCircle(g geometry) (radius float64, isCircle bool) {
	c, isCircle := g.(circle)
	if isCircle {
		return c.radius, true
	}
	return 0, false
}

func main() {
	rect := rectangle{width: 1, height: 4}
	square := rectangle{width: 4, height: 4}
	c := circle{radius: 4}

	fmt.Println("1x4 rectangle efficiency:", shapeEfficiency(rect))
	fmt.Println("square efficiency:", shapeEfficiency(square))
	fmt.Println("circle efficiency:", shapeEfficiency(c))

	fmt.Println(detectCircle(rect))
	fmt.Println(detectCircle(c))
}
